clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -616,11 +622,16 @@ class FindPFWGLambdaFnVisitor
     auto *M = dyn_cast<CXXMethodDecl>(Call->getDirectCallee());
     if (!M || (M->getOverloadedOperator() != OO_Call))
       return true;
-    const int NumPFWGLambdaArgs = 2; // group and lambda obj
+
+    unsigned int NumPFWGLambdaArgs =
+        M->getNumParams() + 1; // group, optional kernel_handler and lambda obj
     if (Call->getNumArgs() != NumPFWGLambdaArgs)
```

Comment:
Is this check still right?  Do we still need to check some of the below tests in the case where the kernel-handler isn't there, or am I missing that the below tests are all related to the kernel_handler?

Reply:
There is an existing test in CodeGenSYCL testing IR when kernel_handler isn't there. I can add a Sema test too if you'd prefer. I didn't add it in this patch since it existing functionality

Reply:
Ah, by 'test' I meant the 'if' statements below in the same function.

Reply:
Oh I see. No I don't think modification is required. The existing code already tested for case without kernel_handler. I just added what was required if kernel_handler is passed.

Reply:
So my concern is that the correct number of args is either getNumParams OR getNumParams +1, right?  Since the 'handler' is optional?

Does this line 620 need to take that into account now?

Reply:
No `getNumParams ` is either 1 or 2 based on if kernel_handler is passed or not. It is the parameters of SYCL kernel.


```
(gdb) p M->dump()
CXXMethodDecl 0xf3b9280 <kernel-handler.cpp:23:42, line:25:9> line:23:9 used operator() 'void (group<1>, sycl::kernel_handler) const' inline
|-ParmVarDecl 0xf3b6ed0 <col:13, col:22> col:22 G 'group<1>':'sycl::group<1>'
|-ParmVarDecl 0xf3b6f48 <col:25, col:40> col:40 kh 'sycl::kernel_handler'
`-CompoundStmt 0xf3b9410 <col:44, line:25:9>
  `-DeclStmt 0xf3b93f8 <line:24:11, col:24>
    `-VarDecl 0xf3b9358 <col:11, col:23> col:15 local 'int' cinit
      `-ImplicitCastExpr 0xf3b93e0 <col:23> 'int' <LValueToRValue>
        `-DeclRefExpr 0xf3b93c0 <col:23> 'const int' lvalue Var 0xf3b2ae0 'a' 'int'
```
 Call is OperatorCall

```
(gdb) p Call->dump()
CXXOperatorCallExpr 0xf3c01c8 'void':'void' '()'
|-ImplicitCastExpr 0xf3c00b8 'void (*)(group<1>, class sycl::kernel_handler) const' <FunctionToPointerDecay>
| `-DeclRefExpr 0xf3c0038 'void (group<1>, class sycl::kernel_handler) const' lvalue CXXMethod 0xf3b9280 'operator()' 'void (group<1>, class sycl::kernel_handler) const'
|-DeclRefExpr 0xf3bfff8 'const class (lambda at kernel-handler.cpp:23:9)':'const class (lambda at kernel-handler.cpp:23:9)' lvalue ParmVar 0xf3bfbd8 'KernelFunc' 'const class (lambda at kernel-handler.cpp:23:9) &'
|-CXXConstructExpr 0xf3c0150 'group<1>':'class sycl::group<1>' 'void (class sycl::group<1> &&) noexcept' elidable
| `-MaterializeTemporaryExpr 0xf3c00f8 'group<1>':'class sycl::group<1>' xvalue
|   `-CXXTemporaryObjectExpr 0xf3a6c58 'group<1>':'class sycl::group<1>' 'void (void) noexcept' zeroing
`-CXXConstructExpr 0xf3c0198 'class sycl::kernel_handler' 'void (const class sycl::kernel_handler &) noexcept'
  `-ImplicitCastExpr 0xf3c0180 'const class sycl::kernel_handler' lvalue <NoOp>
    `-DeclRefExpr 0xf3c0018 'class sycl::kernel_handler' lvalue ParmVar 0xf3bfc40 'kh' 'class sycl::kernel_handler'
```

Reply:
Great, thanks!  Looks like I was confuzzled.

Reply:
Thanks for the review!